### PR TITLE
feat(callbacks): add MemoryUsageCallback for CPU/GPU memory monitoring

### LINKED ISSUE #21150: Memory Usage Monitoring Callback
Hi!
I 'd like to contribute to the keras repository along with one colleague of mine. We had an idea to build a new callback -named MemoryUsageCallback- to to enhance memory monitoring during training. The callback could leverage the "psutil" library for CPU memory usage monitoring and "tf.config.experimental.get_memory_info()" for GPU memory monitoring. We believe this feature would be a valuable addition to Keras, mainly because it will provide detailed insights into resource usage and it will assist users in optimizing training performance and debugging memory-related issues. Should we move forward with this contribution? 
We are eager to hear your feedback!

### PR DESCRIPTION
**Pull request by itsmehotpants**
## Description

This PR implements `keras.callbacks.MemoryUsageCallback`, a backend-agnostic callback that monitors and logs CPU and GPU/accelerator memory usage during `model.fit()`.

Fixes #21150.

---

## Problem

There is currently no built-in way to track memory consumption during training. Users must either implement one-off `LambdaCallback` workarounds or reach for external profiling tools. This makes it hard to:

- Detect memory leaks between epochs
- Tune batch size for a given hardware budget
- Log memory alongside loss/accuracy in `CSVLogger` or `TensorBoard`

---

## Solution

A new `MemoryUsageCallback` that:

1. **Injects memory metrics directly into `logs`** so they flow naturally into `History`, `CSVLogger`, the progress bar, and any other downstream callback — no extra code needed.
2. **Works across all Keras backends** (TensorFlow, PyTorch, JAX) using backend-specific APIs.
3. **Degrades gracefully** — if `psutil` is not installed, CPU memory is skipped; if no GPU/accelerator is detected, GPU memory is skipped. The callback never raises in a CPU-only environment.

### Reported metrics

| Key | Description | Source |
|---|---|---|
| `memory/cpu_used_gb` | Process RSS in GiB | `psutil` (optional dep) |
| `memory/gpu_used_gb` | Device memory in GiB | `tf.config` / `torch.cuda` / `torch.mps` / JAX live-buffers |

### Usage

```python
import keras

# Basic — epoch-level memory logged to History and progress bar
callback = keras.callbacks.MemoryUsageCallback()
history = model.fit(X_train, y_train, epochs=10, callbacks=[callback])
print(history.history["memory/cpu_used_gb"])

# With verbose output
callback = keras.callbacks.MemoryUsageCallback(verbose=1)
# Epoch 1/10  loss: 0.42  memory/cpu_used_gb: 1.2300  memory/gpu_used_gb: 0.8750

# Also log at every batch (useful for detecting batch-level memory growth)
callback = keras.callbacks.MemoryUsageCallback(log_per_batch=True)

# Combine with CSVLogger to persist memory stats to disk
model.fit(X_train, y_train, callbacks=[
    keras.callbacks.MemoryUsageCallback(),
    keras.callbacks.CSVLogger("training.csv"),
])
```

---

## Files changed

| File | Change |
|---|---|
| `keras/src/callbacks/memory_usage_callback.py` | **New** — full implementation |
| `keras/src/callbacks/memory_usage_callback_test.py` | **New** — 16 unit tests |
| `keras/src/callbacks/__init__.py` | Import added |
| `keras/api/callbacks/__init__.py` | Public API export added |

---

## Checklist

- [x] New callback implemented in `keras/src/callbacks/`
- [x] Unit tests covering all methods and all backends (via mocking)
- [x] Public API exported via `keras/api/callbacks/__init__.py`
- [x] Follows existing callback patterns (`@keras_export`, `io_utils.print_msg`)
- [x] No new hard dependencies (psutil is optional)
- [x] Graceful degradation when psutil / GPU is not available

### CODE REVIEW

**Comment by keerthanakadiri** [comment]
Hi @itsmehotpants , Can you kindly sign the CLA? Thank you!
